src/LandIce/problems/LandIce_StokesFOBase.hpp

```diff
@@ -1065,65 +1065,71 @@ constructVelocityEvaluators (PHX::FieldManager<PHAL::AlbanyTraits>& fm0,
-    std::string sideName;
-    
-    for (int i_lcparams=0; i_lcparams<nlcparams; ++i_lcparams)
-    {
-      auto lcparams_i = lcparams.sublist(util::strint("Parameter",i_lcparams));
+      for(size_t coeff_index = 0; coeff_index<numModes; ++coeff_index)
+      {
+        Teuchos::RCP<Teuchos::ParameterList> p = Teuchos::rcp(new Teuchos::ParameterList(util::strint("Coeff", coeff_index)));
+        p->set< Teuchos::RCP<ParamLib> >("Parameter Library", paramLib);
+        const std::string param_name = coeff_names[coeff_index];
+        p->set<std::string>("Parameter Name", param_name);
+        p->set<Teuchos::RCP<Albany::ScalarParameterAccessors<EvalT>>>("Accessors", this->getAccessors()->template at<EvalT>());
+        p->set<const Teuchos::ParameterList*>("Parameters List", &params->sublist("Parameters"));
+        p->set<double>("Default Nominal Value", 0.);
+        Teuchos::RCP<PHAL::SharedParameter<EvalT,PHAL::AlbanyTraits>> ptr_coeff;
+        ptr_coeff = Teuchos::rcp(new PHAL::SharedParameter<EvalT,PHAL::AlbanyTraits>(*p,dl));
+        fm0.template registerEvaluator<EvalT>(ptr_coeff);
+
+        std::string stateName = mode_names[coeff_index];
+        if(onSide && !PHAL::is_field_evaluated<EvalT>(fm0, stateName, dl->side_layouts.at(sideName)->node_scalar)) {
+          Teuchos::RCP<Teuchos::ParameterList> p = Teuchos::rcp(new Teuchos::ParameterList);
+          Albany::StateStruct::MeshFieldEntity entity = Albany::StateStruct::NodalDistParameter;
+          p = stateMgr.registerStateVariable(stateName, dl->side_layouts.at(sideName)->node_scalar, meshSpecs.ebName, true, &entity, "");
+        }
+        if(!onSide && !PHAL::is_field_evaluated<EvalT>(fm0, stateName, dl->node_scalar)) {
+          Teuchos::RCP<Teuchos::ParameterList> p = Teuchos::rcp(new Teuchos::ParameterList);
+          Albany::StateStruct::MeshFieldEntity entity = Albany::StateStruct::NodalDistParameter;
+          p = stateMgr.registerStateVariable(stateName, dl->node_scalar, meshSpecs.ebName, true, &entity, "");
+        }
+      }
 
       Teuchos::RCP<Teuchos::ParameterList> p = Teuchos::rcp(new Teuchos::ParameterList("LCParam"));
       const std::string param_name = lcparams_i.get<std::string>("Name");
-      std::size_t numModes = lcparams_i.get<int>("Number of modes");
 
       p->set<std::string>("Parameter Name", param_name);
       p->set<std::size_t>("Number of modes", numModes);
 
       for (std::size_t i = 0; i < numModes; ++i) {
         Teuchos::RCP<Teuchos::ParameterList> pi = Teuchos::rcp(new Teuchos::ParameterList("Mode"));
-        pi->set<std::string>("Coefficient Name", lcparams_i.sublist(util::strint("Mode",i)).get<std::string>("Coefficient Name"));
-        pi->set<std::string>("Mode Name", lcparams_i.sublist(util::strint("Mode",i)).get<std::string>("Mode Name"));
+        pi->set<std::string>("Coefficient Name", coeff_names[i]);
+        pi->set<std::string>("Mode Name", mode_names[i]);
         p->sublist(util::strint("Mode",i)) = *pi;
```

Comment:
I still think it would be easier to set the array of modes and the array of coeffs names in the parameter list `p`, rather than set N sublists with two strings each...